src/iexamine.cpp

```diff
@@ -3142,45 +3139,106 @@ void iexamine::aggie_plant( Character &you, const tripoint_bub_ms &examp )
+        add_firestarter( it, firestarters, you, examp );
+    }
+
+    const bool has_bionic_firestarter = you.has_bionic( bio_lighter ) &&
+                                        you.enough_power_for( bio_lighter );
+
+    bool use_cbm = false;
+
+    // Query to use a CBM over another firestarter only if the firestarters list isn't empty.
+    if( has_bionic_firestarter && !firestarters.empty() ) {
+        use_cbm = query_yn( _( "Use a CBM to start a fire" ) );
+    } else if( has_bionic_firestarter ) {
+        use_cbm = true;
+    }
+
+    // Try to start the smoking rack/kiln.
+    if( !use_cbm ) {
+        for( auto &firestarter : firestarters ) {
+            item *it = firestarter.second;
+            const use_function *usef = it->type->get_use( "firestarter" );
+            const firestarter_actor *actor = dynamic_cast<const firestarter_actor *>( usef->get_actor_ptr() );
+            you.add_msg_if_player( _( "You attempt to start a fire with your %s…" ), it->tname() );
+            const ret_val<void> can_use = actor->can_use( you, *it, examp );
+            if( can_use.success() ) {
+                const int charges = actor->use( &you, *it, examp ).value_or( 0 );
+                you.use_charges( it->typeId(), charges );
+                return;
+            } else {
+                you.add_msg_if_player( m_bad, can_use.str() );
+            }
+        }
+    } else {
+        you.mod_power_level( -bio_lighter->power_activate );
+        you.mod_moves( -to_moves<int>( 1_seconds ) );
+        firestarter_actor::resolve_firestarter_use( &you, &get_map(), examp,
+                st );
+    }
+}
+
+static units::volume total_volume_at_pos( const tripoint_bub_ms &examp )
+{
+    map &here = get_map();
     map_stack items = here.i_at( examp );
+    units::volume total_volume = 0_ml;
     for( const item &i : items ) {
-        if( i.typeId() == itype_charcoal ) {
-            add_msg( _( "This kiln already contains charcoal." ) );
-            add_msg( _( "Remove it before firing the kiln again." ) );
-            return;
-        } else if( i.made_of_any( kilnable ) ) {
-            fuel_present = true;
-        } else {
-            add_msg( m_bad, _( "This kiln contains %s, which can't be made into charcoal!" ), i.tname( 1,
-                     false ) );
-            return;
-        }
+        total_volume += i.volume();
     }
 
-    if( !fuel_present ) {
-        add_msg( _( "This kiln is empty.  Fill it with wood or bone and try again." ) );
+    return total_volume;
+}
+
+// Highly modified fermenting vat functions
+void iexamine::kiln_empty( Character &you, const tripoint_bub_ms &examp )
+{
+    if( !kiln_prep( you, examp ) ) {
         return;
     }
+
+    pick_firestarter_and_fire( you, examp, firestarter_actor::start_type::KILN );
+}
+
+static void kiln_prep_internal( Character &you, const tripoint_bub_ms &examp,
+                                std::reference_wrapper<int> char_charges_ref )
```

Comment:
you can return `char_chargers` from the function normally, no need for reference wrapper.